File: Emulator/src/main/content/global/skill/summoning/familiar/npc/AbyssalParasiteNPC.java
package content.global.skill.summoning.familiar.npc;

import content.global.skill.summoning.familiar.BurdenBeast;
import content.global.skill.summoning.familiar.Familiar;
import content.global.skill.summoning.familiar.FamiliarSpecial;
import core.game.node.entity.Entity;
import core.game.node.entity.impl.Projectile;
import core.game.node.entity.player.Player;
import core.game.node.item.Item;
import core.game.world.update.flag.context.Animation;
import core.game.world.update.flag.context.Graphics;
import core.plugin.Initializable;
import core.tools.RandomFunction;
import org.rs.consts.NPCs;

/**
 * The type Abyssal parasite npc.
 */
@Initializable
public class AbyssalParasiteNPC extends BurdenBeast {

    @SuppressWarnings("unused")
    private boolean specialMove = false;

    /**
     * Instantiates a new Abyssal parasite npc.
     */
    public AbyssalParasiteNPC() {
        this(null, NPCs.ABYSSAL_PARASITE_6818);
    }

    /**
     * Instantiates a new Abyssal parasite npc.
     *
     * @param owner the owner
     * @param id    the id
     */
    public AbyssalParasiteNPC(Player owner, int id) {
        super(owner, id, 3000, 12035, 1, 7);
    }

    @Override
    public Familiar construct(Player owner, int id) {
        return new AbyssalParasiteNPC(owner, id);
    }

    @Override
    public boolean isAllowed(Player owner, Item item) {
        if (item.getId() != 1436 && item.getId() != 7936) {
            owner.getPacketDispatch().sendMessage("Your familiar can only hold unnoted essence.");
            return false;
        }
        return super.isAllowed(owner, item);
    }

    @Override
    protected boolean specialMove(FamiliarSpecial special) {
        final Entity target = special.getTarget();
        if (!canCombatSpecial(target)) {
            return false;
        }
        faceTemporary(target, 2);
        sendFamiliarHit(target, 7);
        visualize(Animation.create(7672), Graphics.create(1422));
        Projectile.magic(this, target, 1423, 40, 36, 51, 10).send();
        target.getSkills().decrementPrayerPoints(RandomFunction.random(1, 3));
        return true;
    }

    @Override
    public int[] getIds() {
        return new int[]{NPCs.ABYSSAL_PARASITE_6818, NPCs.ABYSSAL_PARASITE_6819};
    }

}

File: server/src/main/content/global/skill/summoning/familiar/npc/AbyssalParasiteNPC.kt
package content.global.skill.summoning.familiar.npc

import content.global.skill.summoning.familiar.BurdenBeast
import content.global.skill.summoning.familiar.Familiar
import content.global.skill.summoning.familiar.FamiliarSpecial
import core.api.sendMessage
import core.game.node.entity.impl.Projectile
import core.game.node.entity.player.Player
import core.game.node.item.Item
import core.game.world.update.flag.context.Animation
import core.game.world.update.flag.context.Graphics
import core.plugin.Initializable
import core.tools.RandomFunction
import org.rs.consts.Items
import org.rs.consts.NPCs

@Initializable
class AbyssalParasiteNPC @JvmOverloads constructor(owner: Player? = null, id: Int = NPCs.ABYSSAL_PARASITE_6818) :
    BurdenBeast(owner, id, 3000, Items.ABYSSAL_PARASITE_POUCH_12035, 1, 7) {

    @Suppress("unused")
    private val specialMove = false

    override fun construct(owner: Player, id: Int): Familiar {
        return AbyssalParasiteNPC(owner, id)
    }

    override fun isAllowed(owner: Player, item: Item): Boolean {
        if (item.id != Items.RUNE_ESSENCE_1436 && item.id != Items.PURE_ESSENCE_7936) {
            sendMessage(owner, "Your familiar can only hold unnoted essence.")
            return false
        }
        return super.isAllowed(owner, item)
    }

    override fun specialMove(special: FamiliarSpecial): Boolean {
        val target = special.target
        if (!canCombatSpecial(target)) {
            return false
        }
        faceTemporary(target, 2)
        sendFamiliarHit(target, 7)
        visualize(Animation.create(7672), Graphics.create(1422))
        Projectile.magic(this, target, 1423, 40, 36, 51, 10).send()
        target.getSkills().decrementPrayerPoints(RandomFunction.random(1, 3).toDouble())
        return true
    }

    override fun getIds(): IntArray {
        return intArrayOf(NPCs.ABYSSAL_PARASITE_6818, NPCs.ABYSSAL_PARASITE_6819)
    }
}
